Mido key implimentation

UnicornJessi opened the pull request:
We are now dual parsing.
Mido parser now uses mido (for midi parsing) and music21 (for key analysis).

Mido parser returns tuple[MidiFile, key.Key] instead of just MidiFile.
All tokenization methods now handle the new tuple format.

julienpoeschl commented on src/groove_panda/midi/parser.py:
```diff
@@ -103,3 +103,27 @@ def _parse_midi_mido(midi_path: str) -> MidiFile:
+        if not isinstance(analyzed_key, key.Key):
+            # Fallback to C major if analysis fails
+            analyzed_key = key.Key("C", "major")
+            logger.warning(f"Key analysis failed for {midi_path}, using C major as fallback")
```
Das Warning könnte man umschreiben, um genau zu sagen, dass analyze keinen key returned hat und deswegen der fallback verwendet wird

julienpoeschl commented on src/groove_panda/processing/tokenization/tokenizer.py:
```diff
@@ -307,7 +307,7 @@ def __init__(self, processed_dataset_id: str = "EMPTY"):
         self.sixtuple_token_maps = SixtupleTokenMaps()
 
-    def tokenize(self, parsed_midi: stream.Score | MidiFile) -> list[Sixtuple]:
+    def tokenize(self, parsed_midi: stream.Score | MidiFile | tuple[MidiFile, key.Key]) -> list[Sixtuple]:
```
Warum ist nur MidiFile noch ein parameter? Das Program sollte ja jetzt nur noch tupel verwenden. _parse_midi_mido_with_key returned ja egal was passiert das tupel

UnicornJessi replied:
vergessen

julienpoeschl commented on src/groove_panda/processing/tokenization/tokenizer.py:
```diff
@@ -323,25 +323,33 @@ def tokenize(self, parsed_midi: stream.Score | MidiFile) -> list[Sixtuple]:
         if isinstance(parsed_midi, stream.Score):
             return self._tokenize_original_key_score(parsed_midi)
 
+        if isinstance(parsed_midi, tuple):
```
Kann man nach `isinstance(parsed_midi, tuple[MidiFile, key.Key])` fragen? Oder geht `isinstance` nur mit `tuple`

UnicornJessi replied:
nope geht nicht isinstance() kann nur mit basis type (tuple, list, dict, str) prüfen

julienpoeschl commented on src/groove_panda/generation/generate.py:
```diff
@@ -1,4 +1,5 @@
 from dataclasses import dataclass
```
HAT NICHTS MIT DEM CODE AN DER STELLE ZU TUN

Wenn ein anderer Tokenize Mode als Original vorliegt, werden die transponierten songs auch einzeln abgespeichert, oder wie ein großer langer song gehandhabt?

UnicornJessi replied:
die werden als ein großer langer song gehandhabt

julienpoeschl replied:
Das ist eher nicht so gut, oder?

UnicornJessi replied:
aber ist es dann nicht auch bei music21 das problem, nicht nur bei mido? weil an sich habe ich ja nichts an _tokenize_all_keys_midi_file() geändert

julienpoeschl replied:
Das war vorher schon ein Problem. Wäre es zu viel verlangt, wenn du noch herausfindest, wie man das stattdessen machen könnte? Vllt wenn statt alle hintereinander, eine liste von listen von sixtuples verwendet wird. Dann könnte die continuous methode von dir die liste der listen erwarten und im normalfall eine liste mit nur einer liste erhalten, und sonst eine liste aus mehreren listen, wo er dann  erkennt, dass das der selbe song aber transponiert ist. Dann könnte man das auch actually nutzen. Ich weiß aber nicht, wie viel man ändern müsste. Hätte nur gedacht: return von tokenize, eingang in die numerize und so, und da wo wir es speichern, dass statt 1 dir n dir mit den n transponierten songs entstehen

UnicornJessi replied:
ye ich schau es mir mal an

UnicornJessi commented:
Fix: Seperating transposed songs in all_keys
-One MIDI file that has 12 transposed versions has now 12 seperate .npz files
and returns 12 seperate lists instead of one
-tokenize() now returns list[list[Sixtuple]] instead of list[Sixtuple]
-added _tokenize_all_keys_as_seperate_list() 
-removed _tokenize_all_keys(), _tokenize_all_keys_score(), _tokenize_all_keys_midi_file()

<img width="135" height="308" alt="Bildschirmfoto 2025-08-06 um 00 41 57" src="https://github.com/user-attachments/assets/91e9bcbd-a647-47d5-8568-394fa498cfc1" />

UnicornJessi commented on src/groove_panda/processing/tokenization/tokenizer.py:
```diff
@@ -307,41 +307,61 @@ def __init__(self, processed_dataset_id: str = "EMPTY"):
-        elif config.tokenize_mode is TokenizeMode.C_MAJOR_A_MINOR:
+            return [sixtuples]
+        if config.tokenize_mode is TokenizeMode.ALL_KEYS:
+            return self._tokenize_all_keys_as_seperate_lists(parsed_midi)
```
uhm not rly 
ich trenn es wieder xD

UnicornJessi replied:
dachte einfach es ist weniger code duplicate
